lib/rubocop/cop/rspec/described_class.rb

```diff
@@ -107,16 +106,84 @@ def skippable_block?(node)
         end
 
         def skip_blocks?
-          cop_config['SkipBlocks'].equal?(true)
+          cop_config['SkipBlocks']
         end
 
         def offensive?(node)
           if style == :described_class
-            node.eql?(@described_class)
+            offensive_described_class?(node)
           else
             node.send_type? && node.method_name == :described_class
           end
         end
+
+        def offensive_described_class?(node)
+          return unless node.const_type?
+
+          nearest_described_class, = node.each_ancestor(:block)
+            .map { |ancestor| described_constant(ancestor) }.find(&:itself)
+
+          return if nearest_described_class.equal?(node)
+
+          full_const_name(nearest_described_class) == full_const_name(node)
+        end
+
+        def full_const_name(node)
+          collapse_namespace(namespace(node), const_name(node))
+        end
+
+        # @param namespace [Array<Symbol>]
+        # @param const [Array<Symbol>]
+        # @return [Array<Symbol>]
+        # @example
+        #   # nil represents base constant
+        #   collapse_namespace([], :C)                 # => [:C]
+        #   collapse_namespace([:A, :B], [:C)          # => [:A, :B, :C]
+        #   collapse_namespace([:A, :B], [:B, :C)      # => [:A, :B, :C]
+        #   collapse_namespace([:A, :B], [nil, :C)     # => [nil, :C]
+        #   collapse_namespace([:A, :B], [nil, :B, :C) # => [nil, :B, :C]
+        def collapse_namespace(namespace, const)
+          return const if namespace.empty?
+          return const if const.first.nil?
+
+          start = [0, (namespace.length - const.length)].max
+          max = namespace.length
+          intersection = (start..max).find do |shift|
+            namespace[shift, max - shift] == const[0, max - shift]
+          end
+          [*namespace[0, intersection], *const]
+        end
+
+        # @param node [RuboCop::AST::Node]
+        # @return [Array<Symbol>]
+        # @example
+        #   const_name(s(:const, nil, :C))                # => [:C]
+        #   const_name(s(:const, s(:const, nil, :M), :C)) # => [:M, :C]
+        #   const_name(s(:const, s(:cbase), :C))          # => [nil, :C]
+        def const_name(node)
+          # rubocop:disable InternalAffairs/NodeDestructuring
+          namespace, name = *node
+          # rubocop:enable InternalAffairs/NodeDestructuring
+          if !namespace
+            [name]
+          elsif namespace.cbase_type?
+            [nil, name]
+          else
+            [*const_name(namespace), name]
+          end
+        end
+
+        # @param node [RuboCop::AST::Node]
+        # @return [Array<Symbol>]
+        # @example
+        #   namespace(node) # => [:A, :B, :C]
+        def namespace(node)
+          node
+            .each_ancestor(:class, :module)
+            .reverse_each
+            .flat_map { |ancestor| ancestor.defined_module_name.split('::') }
+            .map(&:to_sym)
+        end
```

Comment:
It looks as over-engineering to me as well.
I'd rather convert both example group's constant and suspect constant to their fully namespaced variants, and would find and collapse intersections between those strings. However, it doesn't seem to be much simpler than those several methods above.